### DISCUSSION TITLE
Include() in StreamAsync() alternative

### DISCUSSION THREAD

**Discussion by oiBio** [ravendb/ravendb#12399] [Q&A] [answered]
Hello,

we have a quite big query with 13 Include() to dependant documents. 
We wanted to change the query to a StreamResult but unfortunatly includes are not supported and we get the following exception.

**NotSupportedException: System.NotSupportedException: Includes are not supported by this type of query.**

Which is understandable, but can we archieve a similar result just with StreamAsync() in an effective manner?

**Comment by ayende** [comment, marked as answer]
Let's consider this query:

```
from Orders as o 
where o.ShipTo.City = 'London' 
include o.Company, o.Employee
```

When you run this as a query, we get all the results back from the server and store them in the session. Then we can call `Load` on them without going to the server again.

With streaming, however, you are explicitly skipping the session cache and you'll potentially accept a _lot_ of results.
As such, you cannot `include` in the query. Because you'll not be able to run it in this manner, the includes are in the end and you are streaming through the results.

The answer to the problem is that you'll need to change the query, like so:

```
from Orders as o 
where o.ShipTo.City = 'London' 
load o.Company as c, o.Employee as e
select o as Order,e as Employee,c as Company
```

Unlike with a normal query, with streaming, you are not storing in the session cache, so you don't care about the fact that this is a projection or will not be tracked by the session.

In other words, you tell RavenDB that instead of including the details in the end, you'll get them right away.

This has the disadvantage that you'll not be able to collapse multiple references to a single one (as in includes), but the advantage that you need to maintain no state during the streaming process.

**Reply by oiBio** [reply]
Ah okay, that makes sense to me, i just need to return the results directly in my query
Thank you very much for the quick and detailed answer!

**Reply by mcrio** [reply]
Great tip. Could be part of the official documentation.  👍
